src/main/java/io/cdap/plugin/salesforce/SObjectDescriptor.java

```diff
@@ -80,70 +87,184 @@ public static SObjectDescriptor fromQuery(String query) {
   }
 
   public SObjectDescriptor(String name, List<FieldDescriptor> fields) {
+    this(name, fields, Collections.emptyList());
+  }
+
+  public SObjectDescriptor(String name, List<FieldDescriptor> fields, List<SObjectDescriptor> childSObjects) {
     this.name = name;
     this.fields = new ArrayList<>(fields);
+    this.childSObjects = childSObjects;
   }
 
   public String getName() {
     return name;
   }
 
   /**
-   * Collects sObject names needed to be described in order to obtains field type information.
+   * Collects sObject names needed to be described in order to obtain field type information.
+   * Examples:
+   * <ul>
+   *  <li>
+   *    <ol>
+   *      <li>Query: SELECT Id from Opportunity</li>
+   *      <li>Output: [{Opportunity=TOP_LEVEL}]</li>
+   *      <li>Description: `Id` field is simple and does not have parents,
+   *      thus to describe this field (define its type),
+   *      only SObject indicated in the FROM clause is needed,
+   *      i.e. top level SObject: {Opportunity=TOP_LEVEL}.</li>
+   *    </ol>
+   *  </li>
+   *  <li>
+   *    <ol>
+   *      <li>Query: SELECT Id, Account.Name FROM Contact</li>
+   *      <li>Output: [{Contact=TOP_LEVEL}, {Contact=TOP_LEVEL, Account=REFERENCE}]</li>
+   *      <li>Description: `Id` field is simple and does not have parents,
+   *      only top level SObject is needed to define its type ({Contact=TOP_LEVEL}).
+   *      `Name` is reference field, to define its type top level and reference SObjects
+   *      are needed: {Contact=TOP_LEVEL, Account=REFERENCE}.</li>
+   *    </ol>
+   *  </li>
+   *  <li>
+   *    <ol>
+   *      <li>Query: SELECT Name, (SELECT LastName FROM Contacts) FROM Account</li>
+   *      <li>Output: [{Account=TOP_LEVEL}, {Account=TOP_LEVEL, Contacts=RELATIONSHIP}]</li>
```

Comment:
Is the top level object always present in any map returned? And is each map at most 2 items? I'm wondering if this can be more clearly represented as a new class that looks something like:

```
public class QuerySObjects {
  private final String topLevelObject;
  private final Set<String> relationships;
  private final Set<String> references;
}
```

When there are nested collections like a `Set<Map<String, SObjectType>>`, it's hard to understand what each object in each collection represents.